tools/lsp/properties.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+use i_slint_compiler::diagnostics::Spanned;
+use i_slint_compiler::object_tree::{Element, ElementRc};
+
+#[derive(serde::Deserialize, serde::Serialize)]
+pub(crate) struct PropertyInformation {
+    name: String,
+    type_name: String,
+    declared_at: Option<(String, u32)>,
+    defined_at: Option<(u32, u32)>, // Range in the elements source file!
+}
+
+#[derive(serde::Deserialize, serde::Serialize)]
+pub(crate) struct QueryPropertyResponse {
+    properties: Vec<PropertyInformation>,
```

Comment:
My feeling is that the output should resemble the slint files and the data structure in the compiler: Two vectors, one  being a list of available properties (called property declarations) and one with property bindings.

An additional benefit is that this also makes the naming clearer - "PropertyDeclaration" and "PropertyBinding" is clearer than a generic "PropertyInformation".

Reply:
I tried to get as close as possible to what needs to be displayed in the end. That way more code is in the LSP which can is shared between different editor integrations.

Reply:
Can you elaborate on the code sharing part? I don’t understand that.

My points remain: The data structure naming improves and the creation becomes simplified. It feels odd to join two data structures at this level when it is really a feature of the property editor visualization.

Reply:
I want to return something that fits the consumer of the information, even if that makes the producer a bit more complex.

Language servers can be used by lots of editors and I hope many will over time add support for our special features. So I want to make that simple by providing a data response that can be rendered out without too much extra effort in the client of our LS.

I also do not want to expose implementation details of the compiler: A more abstract view on the data seems useful to keep the interface of the LS command stable. The LS interface is public after all, so we need to be careful not to break it too often.

Reply:
I can see how you want to provide an abstract view, instead of letting the language server reflect the language it serves.

I think however that this view is an incomplete abstraction and tailored to only one single consumer.

I think that it is incomplete because it ignores the hierarchy of types and binding overrides within the hierarchy.

I think tailoring what you refer to as abstract view to one single consumer is achieving the opposite. It makes it impossible for example to use the same "view" to write a plugin that detects binding overrides or provides a view to it (a kind of linter).

> I also do not want to expose implementation details of the compiler: A more abstract view on the data seems useful to keep the interface of the LS command stable.

Why is the separation of declaration and binding an implementation detail of the compiler? As far as I can see it is intrinsic to the language and the hierarchy of types, which this data structure does not reflect.

Reply:
Yes, what is in this PR is just a first shot. I fully expect the exposed data and the way it is exposed to change for a while and want to play with an actual UI for this before setting this into stone! But this PR should touch all the corners of the LSP that is needed to get all the information of all the bits and pieces involved (I think;-), so I wanted to get some feedback on whether I forgot something fundamental or did approach the problem from the wrong side:-)

I expect we will end up with something like this in the end:

```rust
struct Property {
   name: String,
   type_name: String,
   declared_at: Option<(file, offset)>, // Maybe leave this out? Mostly so people can jump to declaration
   defined_at: Option<(start_offset, end_offset)>, // No file as this must be in the file that contains the `Element` we get the properties for
   defined_as_value: Option<String>, // Not there right now: The editor can read that from its buffer? This is the expression
   current_value: Option<String>, // The actual calculated value for this property -- if we can get that, which we can not at this time.
   possible_values: Option<Vec<String>>, // Possible values in the case of enums?
};

struct Type {
    name: String, // The type name
    file: Option<String>, // The file this type is defined in
    offset: Option<offset>, // The offset in the file (if any)
    properties: Vec<Property>,
}

struct QueryPropertyResult {
   // Maybe some information about the Element we are looking at (e.g. name, range covered in file, etc.)
   file: String, // The file the Element we are looking at
   properties_by_type: Vec<Type>,
}
```

I do want to have each property listed *once* in one data structure. Otherwise the client authors need to combine e.g. the table of `known properties` with the table of `bindings` (and maybe even more)! I always hated when I had to combine things in Creator: I required me to be aware of all the subtle corner cases that could happen in that particular language.

Reply:
> I wanted to get some feedback on whether I forgot something fundamental or did approach the problem from the wrong side:-)

Yes, feedback given. I agree that fundamentally the LSP should have this type of command that receives a cursor as input and provides information about what's in the surrounding area (depending on the command type).

We disagree on the implementation of how to retrieve the information and how to expose it, but so be it.

Reply:
We do not seem to disagree on the implementation too much anymore :-)

I fully expect the way the information is exposed to keep changing as we try this out -- at least till this lands in an official release. I am in no hurry to merge this to master either!

Reply:
I think I can resolve this now.